nixos/modules/services/misc/persistent-evdev.nix

```
@@ -0,0 +1,66 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.persistent-evdev;
+  settingsFormat = pkgs.formats.json {};
+in
+{
+  options.services.persistent-evdev = {
+    enable = mkEnableOption "virtual input devices that persist even if the backing device is hotplugged";
+
+    devices = mkOption {
+      default = {};
+      type = types.attrsOf types.str;
+      description = ''
+        A set of virtual proxy device labels with backing physical device ids.
+        </para>
+        <para>
+        Physical devices should already exist in <literal>/dev/input/by-id/</literal>.  Proxy
+        devices will be automatically given a <literal>uinput-</literal> prefix.
+        </para>
+        <para>
+        See the <link xlink:href="https://github.com/aiberia/persistent-evdev#example-usage-with-libvirt">
+        project page</link> for example configuration of virtual devices with libvirt
+        and remember to add <literal>uinput-*</literal> devices to the qemu
+        <literal>cgroup_device_acl</literal> list (see <xref linkend="opt-virtualisation.libvirtd.qemu.verbatimConfig"/>).
+      '';
+      example = literalExpression ''
+        {
+          persist-mouse0 = "usb-Logitech_G403_Prodigy_Gaming_Mouse_078738533531-event-if01";
+          persist-mouse1 = "usb-Logitech_G403_Prodigy_Gaming_Mouse_078738533531-event-mouse";
+          persist-mouse2 = "usb-Logitech_G403_Prodigy_Gaming_Mouse_078738533531-if01-event-kbd";
+          persist-keyboard0 = "usb-Microsoft_Natural®_Ergonomic_Keyboard_4000-event-kbd";
+          persist-keyboard1 = "usb-Microsoft_Natural®_Ergonomic_Keyboard_4000-if01-event-kbd";
+        }
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    environment.etc."persistent-evdev.json".source =
+      settingsFormat.generate "persistent-evdev-config" {
+        cache = "/var/cache/persistent-evdev";
+        devices = mapAttrs (virt: phys: "/dev/input/by-id/${phys}") cfg.devices;
+      };
+
+    systemd.services.persistent-evdev = {
+      documentation = [ "https://github.com/aiberia/persistent-evdev/blob/master/README.md" ];
+      description = "Persistent evdev proxy";
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        Restart = "on-failure";
+        ExecStart = "${pkgs.persistent-evdev}/bin/persistent-evdev.py %E/${"persistent-evdev.json"}";
```

```suggestion
        ExecStart = "${pkgs.persistent-evdev}/bin/persistent-evdev.py %E/persistent-evdev.json";
```
although it's generally preferrable to not dump configuration files into `/etc` whenever feasible. is there a utility that needs access to this file at runtime? if not it'd be better to remove the file from /etc (not least because with this unit changes in the config would not restart the service!)